### ISSUE TITLE
Data exports being written to old src path in Azure Blob Storage

### ISSUE THREAD

**Issue by lcjohnso** [zooniverse/panoptes#3539]
**UPDATE: Edits in bold to bring up-to-date with current understanding of issue.**

TL;DR -- Data exports that are requested AND previously existed are **not being uploaded into the correct location**: the CSV media files **at the expected src path remain not updated** but the worker is not throwing any errors and the media entry in the DB is updated such that it appears everything is working correctly.

**Desired Behavior:** When a new data export is requested through the lab (e.g., a workflow classification export) and the export has previously been requested (i.e., there is an existing media item entry in DB + CSV in Azure Blob storage), the dump worker executes the following: update the CSV in blob store with the new dump, the media entry in the DB is updated with a new `updated_at` timestamp + metadata is updated with `"state": "ready"`, and mailer is set to requester to notify them that the dump is complete.

**Current Behavior:** The dump workers seems to exit without errors (no entries show in Honeybadger), and the media entry in the DB is correctly updated, but **the new CSV export file is being stored at the old AWS src path**.

Example: I requested a workflow classification export for workflow_id=7924, which is linked to the following media item: id=63620800, type='workflow_classifications_export', and src='panoptes-uploads.zooniverse.org/production/workflow_classifications_export/7859c7a6-1e6a-46ce-945d-13d3db461f4c.csv'.  The `updated_at` tag is now 2020-12-21, but the file in the blob store is last updated on 2020-11-30 (when the file was migrated from S3) and was not updated.  As a result, the file is missing classifications after 2020-11-11 (when it was last requested; the updated_at value before the most recent unsuccessful request).

If the export has not been previously requested, and the media resource does not exist, then the data export **_does_** seems to work OK.  Example: workflow_id=17025 has export with media id=96429345, type='workflow_classifications_export', and src='workflow_classifications_export/0aae2628-b4d2-45fd-b8f6-0a3666904584.csv' (**NOTE: new simple `src` format!**), and the file in the blob store was last updated on 2020-12-21, in agreement with the media updated_at timestamp.

**Comment by camallen** [comment]
I looked into this today after seeing some unexpected object paths in the azure `panoptesuploads` storage account (`private` container).

What is happening is for older exports that are reusing their medium resource from s3 times. The path that is being uploaded to azure `panoptesuploads` blob storage account in the private container is the old s3 path, i.e. `panoptes-uploads.zooniverse.org/production/workflow_classifications_export/7859c7a6-1e6a-46ce-945d-13d3db461f4c.csv`

So the data is being persisted and the exports are working ok. 

However when you visit the UI to get the download links for the data files, the system is expecting them to be in a different (non s3 prefixed) location i.e. `workflow_classifications_export/7859c7a6-1e6a-46ce-945d-13d3db461f4c.csv` and thus the link we generate for the UI is broken. 

There are two ways we can approach this fix, we can:
1. rewrite the paths before we pass to the azure object store client in like we do for the get URLs so should be pretty easy and avoids updating any db data https://github.com/zooniverse/panoptes/blob/47f7eb0ebd32231da0e9af8dbf55bc19050e4d78/app/services/media_storage/azure_adapter.rb#L63-L70
2. modify the `medium.src` attribute in the dump processor before we attempt the azure object store `put_file` call https://github.com/zooniverse/panoptes/blob/47f7eb0ebd32231da0e9af8dbf55bc19050e4d78/app/models/csv_dumps/dump_processor.rb#L53

I'd lean towards 1 as it may catch some other issues where we are reusing the medium src attribute to `put_files` (do we reuse avatars, project backgrounds etc?). However when doing so we should be careful that there aren't any undesired side effects when putting data so solution 2 may be an option here.

**Comment by nciemniak** [comment]
I discussed this issue with @lcjohnso, and we talked about the possibility of rewriting the actual `src` values of the `medium` records in the database using a rake task. For example, a medium with src "panoptes-uploads.zooniverse.org/production/project_subjects_export/11cd391a-ee2b-4cc1-a8fe-d369c5e717a0.csv" would be updated to "project_subjects_export/11cd391a-ee2b-4cc1-a8fe-d369c5e717a0.csv". Reasoning behind doing this:
- in the case of exports, we dont have to worry about breaking any hard-coded links, since an export can only be retrieved with an SAS signature generated by the panoptes API (valid for 3 minutes)
- Total number of export medium records to be updated is 13709, not a huge number of records to update
- we'd have the option of breaking this out into multiple rake tasks, e.g. one for each export type, allowing us to test it out on a small batch the first time over (e.g. `project_aggregations_export` has less than 200 files I believe) to make sure it works as expected. Also would be easy to test even before rolling out with some test data
- this solution will avoid complicating the code for the sake of solving a problem that we do not expect to resurface (i.e., we dont expect to be switching back to AWS and generating _new_ media records with the panoptes-uploads.zooniverse.org prefix)

**Comment by camallen** [comment]
Makes sense to me 👍

**Comment by nciemniak** [comment]
Notes on steps taken to resolve this can be found [here](https://docs.google.com/document/d/12NPwLfHbXopprBBwnW6Cti3cJp8Y-wEWux1hLEe8IOs/edit)!
